lib/auth/auth_with_roles.go

```diff
@@ -4196,18 +4225,40 @@ func (a *ServerWithRoles) GetRole(ctx context.Context, name string) (types.Role,
 	// Current-user exception: we always allow users to read roles
 	// that they hold.  This requirement is checked first to avoid
 	// misleading denial messages in the logs.
-	if !slices.Contains(a.context.User.GetRoles(), name) {
-		if err := a.action(apidefaults.Namespace, types.KindRole, types.VerbRead); err != nil {
-			return nil, trace.Wrap(err)
-		}
+	if slices.Contains(a.context.User.GetRoles(), name) {
+		return a.authServer.GetRole(ctx, name)
 	}
-	return a.authServer.GetRole(ctx, name)
+
+	authErr := a.action(apidefaults.Namespace, types.KindRole, types.VerbRead)
+	if authErr == nil {
+		return a.authServer.GetRole(ctx, name)
+	}
+
+	// See if the user has access to this individual role.
+	role, err := a.authServer.GetRole(ctx, name)
+	if err != nil {
+		return nil, trace.Wrap(authErr)
+	}
+
+	if err := a.actionForResource(role, types.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return role, nil
 }
 
 // DeleteRole deletes role by name
 func (a *ServerWithRoles) DeleteRole(ctx context.Context, name string) error {
-	if err := a.action(apidefaults.Namespace, types.KindRole, types.VerbDelete); err != nil {
-		return trace.Wrap(err)
+	if authErr := a.action(apidefaults.Namespace, types.KindRole, types.VerbDelete); authErr != nil {
+		// See if the user has access to this individual role.
+		role, err := a.authServer.GetRole(ctx, name)
+		if err != nil {
+			return trace.Wrap(authErr)
+		}
+
+		if err := a.actionForResource(role, types.VerbRead); err != nil {
```

Comment:
VerbDelete here?

I'm also not sure it's valid to only call `actionForResource` if `action` fails, I'm pretty sure there can be cases where `action` returns nil but `actionForResource` would return an error

consider:
```yaml
kind: role
metadata:
  name: noncritical-role-deleter
spec:
  allow:
    rules:
    - resources:
      - role
      verbs: [delete]
      where: '!equals(resource.metadata.labels["critical"], "true")'
```